Test/Microsoft.ApplicationInsights.Test/Shared/Extensibility/Implementation/Endpoints/EndpointProviderTests.cs

```diff
@@ -0,0 +1,84 @@
+        }
+
+        [TestMethod]
+        public void TestEndpointSuffix_WithExplicitOverride()
+        {
+            RunTest(
+                connectionString: "InstrumentationKey=00000000-0000-0000-0000-000000000000;EndpointSuffix=ai.contoso.com;ProfilerEndpoint=https://custom.profiler.contoso.com:444/",
+                expectedBreezeEndpoint: "https://dc.ai.contoso.com/",
+                expectedLiveMetricsEndpoint: "https://live.ai.contoso.com/",
+                expectedProfilerEndpoint: "https://custom.profiler.contoso.com:444/",
+                expectedSnapshotEndpoint: "https://snapshot.ai.contoso.com/"); 
+        }
+
+        [TestMethod]
+        public void TestEndpointSuffix_WithLocation()
+        {
+            RunTest(
+                connectionString: "InstrumentationKey=00000000-0000-0000-0000-000000000000;EndpointSuffix=ai.contoso.com;Location=westus2",
+                expectedBreezeEndpoint: "https://westus2.dc.ai.contoso.com/",
+                expectedLiveMetricsEndpoint: "https://westus2.live.ai.contoso.com/",
+                expectedProfilerEndpoint: "https://westus2.profiler.ai.contoso.com/",
+                expectedSnapshotEndpoint: "https://westus2.snapshot.ai.contoso.com/");
+        }
+
+        [TestMethod]
+        public void TestEndpointSuffix_WithLocation_WithExplicitOverride()
+        {
+            RunTest(
+                connectionString: "InstrumentationKey=00000000-0000-0000-0000-000000000000;EndpointSuffix=ai.contoso.com;Location=westus2;ProfilerEndpoint=https://custom.profiler.contoso.com:444/",
+                expectedBreezeEndpoint: "https://westus2.dc.ai.contoso.com/",
+                expectedLiveMetricsEndpoint: "https://westus2.live.ai.contoso.com/",
+                expectedProfilerEndpoint: "https://custom.profiler.contoso.com:444/",
+                expectedSnapshotEndpoint: "https://westus2.snapshot.ai.contoso.com/");
+        }
+
+        private void RunTest(string connectionString, string expectedBreezeEndpoint, string expectedLiveMetricsEndpoint, string expectedProfilerEndpoint, string expectedSnapshotEndpoint)
+        {
+            var endpoint = new EndpointProvider()
+            {
+                ConnectionString = connectionString
+            };
+
+            var breezeTest = endpoint.GetEndpoint(EndpointName.Breeze);
+            Assert.AreEqual(expectedBreezeEndpoint, breezeTest.AbsoluteUri);
+
+            var liveMetricsTest = endpoint.GetEndpoint(EndpointName.LiveMetrics);
+            Assert.AreEqual(expectedLiveMetricsEndpoint, liveMetricsTest.AbsoluteUri);
+
+            var profilerTest = endpoint.GetEndpoint(EndpointName.Profiler);
+            Assert.AreEqual(expectedProfilerEndpoint, profilerTest.AbsoluteUri);
+
+            var snapshotTest = endpoint.GetEndpoint(EndpointName.Snapshot);
+            Assert.AreEqual(expectedSnapshotEndpoint, snapshotTest.AbsoluteUri);
+        }
+    }
+}
```

Comment:
Would be wise to test error paths too (badly-formed connection strings). e.g. no InstrumentationKey, duplicate keys, illegal characters.

Reply:
Gladly :)
I've got a few already. I'm going to sync up with the other SDK devs and share test cases.